frame/two-phase-election-provider/src/macros.rs

```diff
@@ -0,0 +1,165 @@
+		};
+	};
+}
+
+#[macro_export]
+macro_rules! voter_index_generate_cache {
+	($name:ident, $voters:ident, $t:ident) => {
+		let mut $name:
+			$crate::sp_std::collections::btree_map::BTreeMap<
+				<$t as frame_system::Config>::AccountId,
+				usize
+			>
+			= $crate::sp_std::collections::btree_map::BTreeMap::new();
+		$voters.iter().enumerate().for_each(|(i, (x, _, _))| {
+			let _existed = $name.insert(x.clone(), i);
+			// if a duplicate exists, we only consider the last one. Defensive only, should never happen.
+			debug_assert!(_existed.is_none());
+		});
+	};
+}
+
+#[macro_export]
+macro_rules! voter_index_fn {
+	(let $name:ident, $voters:ident, $t:ident) => {
+		voter_index_generate_cache!(voters_map, $voters, $t);
+		let $name = |who: &<$t as frame_system::Config>::AccountId| -> Option<CompactVoterIndexOf<$t>> {
+			voters_map
+				.get(who)
+				.and_then(|i| <usize as $crate::sp_std::convert::TryInto<CompactVoterIndexOf<$t>>>::try_into(*i).ok())
+		};
+	};
+}
+
+#[macro_export]
+macro_rules! voter_index_fn_usize {
+	(let $name:ident, $voters:ident, $t:ident) => {
+		voter_index_generate_cache!(voters_map, $voters, $t);
+						let $name = |who: &<$t as frame_system::Config>::AccountId| -> Option<usize> {
+							voters_map.get(who).cloned()
+						};
+	};
+}
+
+#[macro_export]
+macro_rules! target_index_fn {
+	(let $name: ident, $targets:ident, $t:ident) => {
+		let $name = |who: &<$t as frame_system::Config>::AccountId| -> Option<$crate::CompactTargetIndexOf<$t>> {
+			$targets
+				.iter()
+				.position(|x| x == who)
+				.and_then(|i|
+					<
+						usize
+						as
+						$crate::sp_std::convert::TryInto<$crate::CompactTargetIndexOf<$t>>
+					>::try_into(i).ok()
+				)
+		};
+	};
+}
+
+#[macro_export]
+macro_rules! voter_at_fn {
+	(let $name:ident, $snap:ident, $t:ident) => {
+		let $name = |i: $crate::CompactVoterIndexOf<$t>| -> Option<<$t as frame_system::Config>::AccountId> {
+			<$crate::CompactVoterIndexOf<$t> as $crate::sp_std::convert::TryInto<usize>>::try_into(i)
+				.ok()
+				.and_then(|i| $snap
+					.get(i)
+					.map(|(x, _, _)| x)
+					.cloned()
+				)
+		};
+	};
+}
+
+#[macro_export]
+macro_rules! target_at_fn {
+	(let $name:ident, $snap:ident, $t:ident) => {
+		let $name = |i: $crate::CompactTargetIndexOf<$t>| -> Option<<$t as frame_system::Config>::AccountId> {
+			<
+				$crate::CompactTargetIndexOf<$t>
+				as
+				$crate::sp_std::convert::TryInto<usize>
+			>::try_into(i)
+				.ok()
+				.and_then(|i| $snap.get(i).cloned())
+		};
+	};
+}
+
+#[deprecated = "May only be used for benchmarking purposes."]
+#[macro_export]
+macro_rules! stake_of_fn_linear {
+	(let $name:ident, $voters:ident, $t:ty) => {
+		let $name = |who: &<$t as frame_system::Config>::AccountId| -> $crate::VoteWeight {
+							$voters
+								.iter()
+								.find(|(x, _, _)| x == who)
+								.map(|(_, x, _)| *x)
+								.unwrap_or_default()
+						};
+	};
+}
```

Comment:
I don't think it is a correct usage of `deprecated`, maybe you can add a suffix `_unoptimized` or hide it behind a feature.